### ISSUE TITLE
Different date results using different python versions (3.4 & 3.5)

### ISSUE THREAD

**Issue by Veebers** [dateutil/dateutil#158]
The following code results in a different datetime results depending on which version of python is used to run it:

``` python
import datetime
from dateutil import tz

tzfile = tz.tzfile('/usr/share/zoneinfo/Europe/London')
print(datetime.datetime.fromtimestamp(1090123200, tzfile))

# Output:
# $ python3.4 example.py
# >>> 2004-07-18 04:00:00+01:00
#
# $ python3.5 example.py
# >>> 2004-07-18 05:00:00+01:00
```

Also note that the result disagrees when using pytz (with either python version):

``` python
import pytz

london = pytz.timezone('Europe/London')
print(datetime.datetime.fromtimestamp(1090123200, london))

# >>> 2004-07-18 05:00:00+01:00
```

**Comment by pganssle** [comment]
@Veebers Which is the correct answer? I get 04:00 on 3.5:

```
Python 3.5.0 |Anaconda 2.4.0 (64-bit)| (default, Oct 19 2015, 21:57:25) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import datetime
>>> from dateutil import tz
>>> 
>>> tzfile = tz.tzfile('/usr/share/zoneinfo/Europe/London')
>>> print(datetime.datetime.fromtimestamp(1090123200, tzfile))
2004-07-18 04:00:00+01:00
```

**Comment by sdiemer** [comment]
Hello,

I think I have the same issue using Ubuntu 14.04 64bits + Python 3.4.3 (system default):

`
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
Type "copyright", "credits" or "license" for more information.

IPython 1.2.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import datetime

In [2]: import dateutil

In [3]: dateutil.__version__
Out[3]: '2.5.2'

In [4]: from dateutil import tz

In [5]: london = tz.gettz('Europe/London')

In [6]: now = datetime.datetime.now(tz.tzutc())

In [7]: print(now)
2016-04-05 12:35:37.170199+00:00

In [8]: print(now.astimezone(london))
2016-04-05 12:35:37.170199+01:00
`

The last print should be "2016-04-05 13:35:37.170199+01:00".
I did not find anything to avoid this problem for now...

**Comment by pganssle** [comment]
@sdiemer Bug is confirmed on Python 3.4 in Ubuntu. Works on 2.7 and 3.5, so maybe in the mean time try upgrading Python to 3.5 if possible. I'll try and figure out what exactly is going on.

**Comment by sdiemer** [comment]
@pganssle For now I cannot update Python, I will get Python 3.5 when Ubuntu 16.04 will be released.
Thanks for your answer.

**Comment by pganssle** [comment]
@sdiemer OK. Now that I can clearly reproduce this, I should be able to track it down. I may be able to come up with a mitigation, but either way I'll try to fix it before the 2.5.3 release.

**Comment by pganssle** [comment]
Wow, this is bizarre. I cut-and-pasted the `fromutc()` function from the Python 3.4.3 source distribution into `tzfile()` to see where it was going wrong (this is, mind you, not a function that has changed whatsoever between Python 3.3 and 3.5), and it fixed the issue.

I'm thinking this may be something extremely specific to the Ubuntu release somehow. I don't think it's something I can (or rather, should) fix on the dateutil end, unless I'm missing something.

For now, here's a workaround that patches the correct `fromutc` function into `tzfile`:

```python
import datetime
from dateutil import tz

class tzinfo_patch(object):
    def fromutc(self, dt):
        "datetime in UTC -> datetime in local time."

        if not isinstance(dt, datetime.datetime):
            raise TypeError("fromutc() requires a datetime argument")
        if dt.tzinfo is not self:
            raise ValueError("dt.tzinfo is not self")

        dtoff = dt.utcoffset()
        if dtoff is None:
            raise ValueError("fromutc() requires a non-None utcoffset() "
                             "result")

        # See the long comment block at the end of this file for an
        # explanation of this algorithm.
        dtdst = dt.dst()
        if dtdst is None:
            raise ValueError("fromutc() requires a non-None dst() result")
        delta = dtoff - dtdst
        if delta:
            dt += delta
            dtdst = dt.dst()
            if dtdst is None:
                raise ValueError("fromutc(): dt.dst gave inconsistent "
                                 "results; cannot convert")
        return dt + dtdst

class patched_tzfile(tzinfo_patch, tz.tzfile):
    pass

def _patched_gettz(*args, **kwargs):
    tzi = tz.gettz(*args, **kwargs)

    if isinstance(tzi, tz.tzfile):
        return patched_tzfile(tzi._filename)

gettz = _patched_gettz

LONDON = gettz('Europe/London')
UTC = tz.tzutc()

dt1 = datetime.datetime(2016, 4, 5, 12, 35, tzinfo=UTC)
dt_l = dt1.astimezone(LONDON)

print(dt1)          # 2016-04-05 12:35:00+00:00
print(dt_l)         # 2016-04-05 13:35:00+01:00
```

It does rely on a private member function, and I haven't tested it thoroughly, so it might be fragile, but hopefully it will help in the short term if you are stuck on a broken version of Python.

I will try and track down the python source code that Ubuntu used for their 3.4.3 build to see if they changed something and if there's anything I can / should do about it.

The reason that `pytz` works, by the way, is that they use their own implementation of `fromutc`. We may switch to our own implementation in dateutil-2.6.0 if necessary for handling edge cases, in which case this issue will be fixed without upgrading. However, I don't want to duplicate the code from Python just because one particular build had an issue with it.

**Comment by sdiemer** [comment]
Thank you for spending time to find a solution.
I will wait for the Ubuntu 16.04 release. For now, I said to the users having the issue in our software to use a static timezone like GMT+1 to avoid the problem.

**Comment by pganssle** [comment]
Hm.. This is very strange. I spun up a Windows version of Python 3.4.3 and it actually has the same issue. A similar patch fixes it. Maybe there's something I'm missing here.

**Comment by pganssle** [comment]
Ah, I'm fairly certain that this is a bug in Python <= 3.4. See [here](https://bugs.python.org/issue23600). The reason seems to be that the C implementation gives a different result from the Python implementation.

Turns out it only applies during summer time for zones whose normal offset is 0. It should be fixed in version 3.4.4 if that's available to you (at least according to [the NEWS file](https://docs.python.org/3.4/whatsnew/changelog.html#python-3-4-4)).

I don't think that I will deliberately fix this bug, but in addition to a possible custom implementation of `fromutc()` (which we'll do only if necessary to deal with edge cases), I would also like to make it easier to construct a custom subclass of `tzfile` objects, so that if people want to extend their functionality (or apply a patch to work around version incompatibility issues), they are free to do that.
